substrate/client/keystore/src/local.rs

```diff
@@ -334,67 +349,89 @@ impl Keystore for LocalKeystore {
-			key_type: KeyTypeId,
-			seed: Option<&str>,
-		) -> std::result::Result<bls381::Public, TraitError> {
-			self.generate_new::<bls381::Pair>(key_type, seed)
-		}
+			/// Generate a new pair compatible with the 'bls381' signature scheme.
+			///
+			/// If `[seed]` is `Some` then the key will be ephemeral and stored in memory.
+			fn bls381_generate_new(
+				&self,
+				key_type: KeyTypeId,
+				seed: Option<&str>,
+			) -> std::result::Result<bls381::Public, TraitError> {
+				self.generate_new::<bls381::Pair>(key_type, seed)
+			}
 
-		fn bls381_sign(
-			&self,
-			key_type: KeyTypeId,
-			public: &bls381::Public,
-			msg: &[u8],
-		) -> std::result::Result<Option<bls381::Signature>, TraitError> {
-			self.sign::<bls381::Pair>(key_type, public, msg)
-		}
+			fn bls381_sign(
+				&self,
+				key_type: KeyTypeId,
+				public: &bls381::Public,
+				msg: &[u8],
+			) -> std::result::Result<Option<bls381::Signature>, TraitError> {
+				self.sign::<bls381::Pair>(key_type, public, msg)
+			}
 
-		fn ecdsa_bls381_public_keys(&self, key_type: KeyTypeId) -> Vec<ecdsa_bls381::Public> {
-			self.public_keys::<ecdsa_bls381::Pair>(key_type)
-		}
+			fn bls381_generate_pop(
+				&self,
+				key_type: KeyTypeId,
+				public: &bls381::Public
+			) -> std::result::Result<Option<bls381::Signature>, TraitError> {
+				self.generate_pop::<bls381::Pair>(key_type, public)
+			}
 
-		/// Generate a new pair of paired-keys compatible with the '(ecdsa,bls381)' signature scheme.
-		///
-		/// If `[seed]` is `Some` then the key will be ephemeral and stored in memory.
-		fn ecdsa_bls381_generate_new(
-			&self,
-			key_type: KeyTypeId,
-			seed: Option<&str>,
-		) -> std::result::Result<ecdsa_bls381::Public, TraitError> {
-			self.generate_new::<ecdsa_bls381::Pair>(key_type, seed)
-		}
+			fn ecdsa_bls381_public_keys(&self, key_type: KeyTypeId) -> Vec<ecdsa_bls381::Public> {
+				self.public_keys::<ecdsa_bls381::Pair>(key_type)
+			}
 
-		fn ecdsa_bls381_sign(
-			&self,
-			key_type: KeyTypeId,
-			public: &ecdsa_bls381::Public,
-			msg: &[u8],
-		) -> std::result::Result<Option<ecdsa_bls381::Signature>, TraitError> {
-			self.sign::<ecdsa_bls381::Pair>(key_type, public, msg)
-		}
+			/// Generate a new pair of paired-keys compatible with the '(ecdsa,bls381)' signature scheme.
+			///
+			/// If `[seed]` is `Some` then the key will be ephemeral and stored in memory.
+			fn ecdsa_bls381_generate_new(
+				&self,
+				key_type: KeyTypeId,
+				seed: Option<&str>,
+			) -> std::result::Result<ecdsa_bls381::Public, TraitError> {
+				let pubkey = self.generate_new::<ecdsa_bls381::Pair>(key_type, seed)?;
+
+				let s = self
+					.0
+					.read()
+					.additional
+					.get(&(key_type, pubkey.to_vec()))
+					.map(|s| s.to_string())
+					.expect("Can retrieve seed");
+
+	// This is done to give the keystore access to individual keys, this is necessary to avoid unnecessary host functions for paired keys and re-use host functions implemented for each element of the pair.
```

Comment:
Please split this long comment.
- Lines length at most ~100
- Align first column to the code (format is wrong)